### ISSUE TITLE
[css-color] Clarification needed on how missing alpha components flow through relative color syntax for hsl/hwb

### ISSUE THREAD

**Issue by weinig** [w3c/csswg-drafts#10254]
In the WPT test, [`css/css-color/parsing/color-computed-relative-color.html`](https://github.com/web-platform-tests/wpt/blob/master/css/css-color/parsing/color-computed-relative-color.html), there are a few examples I don't quite understand and can't figure out which language in the spec dictate them.

One is:

```
fuzzy_test_computed_color(`hsl(from rebeccapurple none none none / none)`, `color(srgb 0 0 0 / none)`)
```

What allows the alpha value to be passed through as `none`?

My understand here of the operations performed is:

1. `rebeccapurple` converted to hsl() [`hsl(270 50% 40%)`]
2.  Relative color is resolved via channel arguments to `hsl(none none none / none)`
3. `hsl(none none none / none)` is converted to `color(srgb)` for serialization, resolving all missing components to 0 [`color(srgb 0 0 0 / 0)`] (https://drafts.csswg.org/css-color-5/#serial-relative-color)

Is the intention for that last conversion (from `hsl()` to `color(srgb)`) to carry forward analogous component missing values?

Somewhat perplexingly (and perhaps just an error in the test), the corresponding test [`css/css-color/parsing/color-valid-relative-color`](https://github.com/web-platform-tests/wpt/blob/master/css/css-color/parsing/color-valid-relative-color.html) has a similar example.

```
fuzzy_test_valid_color(`${hslFunction}(from rebeccapurple none none none / none)`, `color(srgb 0 0 0 / 0)`);
```

My expectation is that they should at least expect the same result.

Similarly, what would expect for missing components in the origin.

```css
color: hsl(from hsl(none none none / none) h s l / alpha);
```

```css
color: hsl(from hsl(none none none / none) h s l);
```

**Comment by svgeesus** [comment]
I think this is from [4.1. Processing Model for Relative Colors ](https://drafts.csswg.org/css-color-5/#rcs-intro):

> If the alpha value of the relative color is omitted, it defaults to that of the [origin color](https://drafts.csswg.org/css-color-5/#origin-color) (rather than defaulting to 100%, as it does in the absolute syntax).

> Missing components are handled the same way as with [CSS Color 4 § 12.2 Interpolating with Missing Components](https://drafts.csswg.org/css-color-4/#interpolation-missing): the origin colorspace and the relative function colorspace are checked for [analogous components](https://drafts.csswg.org/css-color-4/#analogous-components) which are then [carried forward](https://drafts.csswg.org/css-color-4/#carried-forward) as missing.

Aha! It is implied, but not directly stated, that alpha is analogous to alpha.

**Comment by svgeesus** [comment]
> Is the intention for that last conversion (from `hsl()` to `color(srgb)`) to carry forward analogous component missing values?

I'm not fully sure on that point.

**Comment by weinig** [comment]
I totally agree that adding something to [12.2. Interpolating with Missing Components](https://drafts.csswg.org/css-color-4/#analogous-components) that says alpha is analogous to alpha would be great.

To restate what is probably obvious though, the bigger issue on this one is should the conversion from HSL()/HWB() to color(srgb) done for serialization be a "normal conversion" or should it be a "conversion that carries forward analogous component missing values"?

**Comment by weinig** [comment]
I thought a bit more about this and I think preserving the NaN (that is, using the "carries forward analogous component missing values" conversion) for alpha is better than not.

My argument is there really isn't any value in converting it to 0 here. As much as possible, serialization tries to preserve as much information as possible, and since we can preserve this, why not. (It also seems to be what Chrome does).

One things this does is beg the question, should all conversions between color spaces carry forward analogous missing components? In what cases would it not be what you want?

**Comment by weinig** [comment]
> Is there any part of the specification that indicates that you shouldn't always carry forward analogous missing components during conversions between color spaces?
> 
> 
> 
> I thought this was always required.

Section 4.4 is pretty explicit. 

> For all other purposes, a missing component behaves as a zero value, in the appropriate unit for that component: 0, 0%, or 0deg. This includes rendering the color directly, converting it to another color space, performing computations on the color component values, etc.

https://www.w3.org/TR/css-color-4/#missing

**Comment by weinig** [comment]
And since interpolation calls it out as a separate step, I think the implication is that if an algorithm wants carrying forward, it needs to call it out.

**Comment by romainmenke** [comment]
Ah, indeed!

The intention is to preserve missing components / `none` as much as possible.

So if there are algorithms where it isn't stated and where it could be preserved, then we should change that.

The serialization as `color(srgb)` for specific cases is problematic when trying to preserve missing components / `none`.

It doesn't currently state that carry forward should happen. But it also makes carry forward impossible for non-alpha components when the src and dest aren't both in srgb to begin with.

**Comment by romainmenke** [comment]
> If you agree, then I can add alpha to [12.2. Interpolating with Missing Components](https://drafts.csswg.org/css-color-4/#analogous-components)

https://codepen.io/romainmenke/pen/LYvvNrb

This would also affect interpolation but there isn't any interop for `none` alpha.
Chrome threats `none` in alpha during interpolation as the alpha of the other color. Safari and Firefox do not.

**Comment by emilio** [comment]
> > If you agree, then I can add alpha to [12.2. Interpolating with Missing Components](https://drafts.csswg.org/css-color-4/#analogous-components)
> 
> https://codepen.io/romainmenke/pen/LYvvNrb
> 
> This would also affect interpolation but there isn't any interop for `none` alpha. Chrome threats `none` in alpha during interpolation as the alpha of the other color. Safari and Firefox do not.

That looks like a bug, if you change it to:

```
 background-color: color-mix(
    in srgb,
    color(srgb 0.4 0.4 0 / none),
    color(srgb 0.4 0 0.4 / 0.5)
  )
```

Then Firefox does the right thing too. It seems we lose the `none` information when converting the colors for interpolation.

**Comment by emilio** [comment]
Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1901991 for this btw, Safari seems to behave like Firefox.

**Comment by emilio** [comment]
Ok, so the issue is that we treat those colors like "legacy" syntax, which generally doesn't allow `none`, but the `/` syntax does.

Chrome's handling seems broken as well, the `none` is completely missing when serializing:

```
a.style.backgroundColor = "rgb(100 100 0 / none)";
a.style.backgroundColor // 'rgba(100, 100, 0, 0)'
```

So, how should those colors serialize? Should we just serialize to `color(srgb ..)`? That'd be preferable IMO...

**Comment by svgeesus** [comment]
> So, how should those colors serialize? Should we just serialize to `color(srgb ..)`? That'd be preferable IMO...

That seems like a good way forward to me

**Comment by svgeesus** [comment]
Agenda+F2F to try and get the feedback on the question asked last year.

**Comment by svgeesus** [comment]
Proposed resolution: missing alpha values are preserved during serialization, and thus serialize in modern syntax (which allows `none`)

```
a.style.backgroundColor = "rgb(100 100 0 / none)";
a.style.backgroundColor // 'color(srgb 1 1 0 /  none)'
```

**Comment by svgeesus** [comment]
This affects `rgb()`, `hsl()` because currently these serialize in legacy syntax "for historical reasons". Should also affect `hwb()`

**Comment by weinig** [comment]
It occurred to me after the discussion that the proposal is very close (if not identical) to what we have for relative color syntax origin color serialization (https://drafts.csswg.org/css-color-5/#serial-origin-color), including preserving hsl()/hwb(). Re-using some of that language will probably make specifying this easier.

**Comment by cdoublev** [comment]
Based on the last changes, I do not know how `rgb(from red 0 none 0)` should serialize as a component of a declared value.
